airflow-core/src/airflow/utils/log/log_reader.py

```diff
@@ -96,10 +99,17 @@ def read_log_stream(self, ti: TaskInstance, try_number: int | None, metadata: di
                 TaskInstanceState.RUNNING,
                 TaskInstanceState.DEFERRED,
             ):
-                if not logs[0]:
+                if logs:
+                    empty_iterations = 0
+                else:
                     # we did not receive any logs in this loop
                     # sleeping to conserve resources / limit requests on external services
                     time.sleep(self.STREAM_LOOP_SLEEP_SECONDS)
+                    empty_iterations += 1
+                    if empty_iterations >= self.STREAM_LOOP_STOP_AFTER_EMPTY_ITERATIONS:
+                        # we have not received any logs for a while, so we stop the stream
+                        yield "(Log stream stopped - End of log marker not found, so logs may be incomplete)\n"
```

Comment:
```suggestion
                        yield "Log stream stopped but no end of log marker not found; logs may be incomplete."
```

Is it supposed to include a newline as you have done?

Reply:
(seems odd if so)

Reply:
We add a newline onto each other line, including the last one, so kept it that way for consistency.

Reply:
I didn't take your message changes as-is, but tried to find a decent middle ground. Take a peek.